diff --git a/Cython/Build/Dependencies.py b/Cython/Build/Dependencies.py
index a031d60..62177c7 100644
--- a/Cython/Build/Dependencies.py
+++ b/Cython/Build/Dependencies.py
@@ -473,7 +473,7 @@ def create_extension_list(patterns, exclude=[], ctx=None, aliases=None):
                 if template is not None:
                     sources += template.sources[1:]
                 module_list.append(exn_type(
-                        name=module_name,
+                        name=str(module_name),
                         sources=sources,
                         **kwds))
                 m = module_list[-1]
